File: app/schemas/billing.py
#The object that saves data individually to send it to backend structuredly.
from pydantic import BaseModel, Field
from typing import Optional, Literal
from datetime import datetime

class BillingCreate(BaseModel):
    farmer_id: str = Field(..., description="MongoDB ID of the farmer as string")
    operator_id: str = Field(..., description="MongoDB ID of the operator/user as string")
    drone_id: str = Field(..., description="MongoDB ID of the drone as string")
    acres: float = Field(..., gt=0)
    time: float = Field(..., gt=0, description="Time in hours")
    amount: float = Field(..., gt=0)
    mode_type: Literal["cash", "upi"]

class BillingUpdate(BaseModel):
    acres: Optional[float] = Field(None, gt=0)
    time: Optional[float] = Field(None, gt=0)
    amount: Optional[float] = Field(None, gt=0)
    mode_type: Optional[Literal["cash", "upi"]] = None

class BillingResponse(BaseModel):
    id: str
    farmer_id: str
    operator_id: str
    drone_id: str
    acres: float
    time: float
    amount: float
    mode_type: Literal["cash", "upi"]
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    created_by: Optional[str] = None
    updated_by: Optional[str] = None

File: frontend/src/apis/billing.ts
import { client } from './client';

export interface BillingResponse {
    _id: string;
    id?: string;
    farmer_id: string;
    farmer_name?: string;
    farmer_number?: string;
    operator_id: string;
    operator_name?: string;
    drone_id: string;
    acres: number;
    time: number;
    amount: number;
    mode_type: 'cash' | 'upi';
    created_at: string;
    updated_at: string | null;
    created_by: string | null;
}

export interface BillingCreate {
    farmer_id: string;
    operator_id?: string;
    drone_id?: string;
    acres?: number;
    time?: number | string;
    amount: number;
    mode_type: 'cash' | 'upi';
}

function normalizeBill(bill: any): BillingResponse {
    if (!bill) return bill;
    const resolvedId = bill._id || bill.id;
    return {
        ...bill,
        _id: resolvedId,
        id: resolvedId,
    };
}

export const billsApi = {
    create: async (data: BillingCreate): Promise<BillingResponse> => {
        const response = await client<BillingResponse>('/billing/', { body: data });
        return normalizeBill(response);
    },

    // Get all bills (with optional filtering)
    getAll: async (filters?: { farmer_id?: string; operator_id?: string; drone_id?: string }): Promise<BillingResponse[]> => {
        const params = new URLSearchParams();
        if (filters) {
            if (filters.farmer_id) params.append('farmer_id', filters.farmer_id);
            if (filters.operator_id) params.append('operator_id', filters.operator_id);
            if (filters.drone_id) params.append('drone_id', filters.drone_id);
        }
        const queryString = params.toString();
        const response = await client<BillingResponse[]>(`/billing/${queryString ? `?${queryString}` : ''}`);
        return (response || []).map(normalizeBill);
    },

    getById: async (id: string): Promise<BillingResponse> => {
        const response = await client<BillingResponse>(`/billing/${id}`);
        return normalizeBill(response);
    },

    update: (id: string, data: Partial<BillingCreate>) => client(`/billing/${id}`, {
        method: 'PUT',
        body: data
    }),

    delete: (id: string) => client(`/billing/${id}`, { method: 'DELETE' }),
};
